Python/Product/PythonTools/PythonTools/Editor/PythonTextBufferInfo.cs

```diff
@@ -288,6 +308,31 @@ sealed class PythonTextBufferInfo {
             }
         }
 
+        /// <summary>
+        /// Returns the current analysis entry if it is not null. Otherwise
+        /// waits for a non-null entry to be set and returns it. If cancelled,
+        /// return null.
+        /// </summary>
+        public Task<AnalysisEntry> GetAnalysisEntryAsync(CancellationToken cancellationToken) {
+            var entry = AnalysisEntry;
+            if (entry != null) {
+                return Task.FromResult(entry);
+            }
+            var tcs = Volatile.Read(ref _waitingForEntry);
+            if (tcs != null) {
+                return tcs.Task;
+            }
+            tcs = new TaskCompletionSource<AnalysisEntry>();
+            tcs = Interlocked.CompareExchange(ref _waitingForEntry, tcs, null) ?? tcs;
+            entry = AnalysisEntry;
+            if (entry != null) {
+                tcs.TrySetResult(entry);
+            } else if (cancellationToken.CanBeCanceled) {
+                cancellationToken.Register(() => tcs.TrySetResult(null));
```

Comment:
`TrySetCanceled`?

Reply:
By design. You need to check for null anyway, so no point forcing all the callers to also check for exceptions and treat them identically to null.

Reply:
That's a bit unusual. For every awaited async method that accepts `cancellationToken`, caller expects that the code below canceled await won't be executed, no matter how deep in the call stack the actual handler of cancellation is placed. 
Here it is actually not true. While immediate caller can check for `null`, other methods in call stack has no idea about it.
Right now, none of the callers provide real `cancellationToken`, but it may change in future.